### ISSUE TITLE
Performance degradation between 21.3 and 22.3

### ISSUE THREAD

**Issue by chen9t** [ClickHouse/ClickHouse#36532]
We are upgrading the ClickHouse cluster from 21.3 to 22.3, and found some performance degradation here.
So we build a clean test environment, which can reproduce this degradation.

### Test environment
4 same machines, 2 for 21.3.13.9 and 2 for 22.3.3.44. And deploy 2 clusters with 2 shards and 0 replica.

### How to reproduce
Create table as standard SSB test from [here](https://clickhouse.com/docs/en/getting-started/example-datasets/star-schema/#), with data distributed to 2 shards for each cluster.

Run `clickhouse-benchmark -r on --json result.json -i 5000 < dist_ssb_queries` on 2 clusters.

Queries in `dist_ssb_queries` are all 13 SSB test queries for distributed table.

### Result
query_time_percentiles | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 16.19 | 20.40 | 22.83 | 68.78 | 161.13 | 181.10 | 190.94 | 261.25 | 273.65 | 300.02 | 320.80 | 329.96 | 337.44 | 353.47
22.3.3.44 | 19.57 | 25.57 | 29.37 | 94.72 | 191.79 | 202.76 | 213.08 | 271.75 | 357.13 | 398.58 | 410.79 | 443.40 | 494.78 | 513.98

And this result can be reproduced stably.

**Comment by chen9t** [comment]
Not sure if it's caused by issues mentioned by #36411, @[nikitamikhaylov](https://github.com/nikitamikhaylov) can you take a look?

Thanks!

**Comment by den-crane** [comment]
>4 same machines, 2 for 21.3.13.9 and 2 for 22.3.3.44. And deploy 2 clusters with 2 shards and 0 replica.

Hardware? CPU / RAM / Disks?

>Create table as standard SSB test from

Star schema does not have Distributed tables. How you created Distrbuted tables? Sharding expression? How did you load data?

>dist_ssb_queries

What is `dist_ssb_queries` ? Which query made the biggest contribution to the slowdown?

**Comment by baibaichen** [comment]
@chen9t  I think it is highly related with decompression degradation.

**Comment by chen9t** [comment]
@den-crane  Sorry I missed some info before, updated and added test result for non-distributed table.

## Hardware
4 [Tencent cloud IT5.16XLARGE256](https://intl.cloud.tencent.com/document/product/213/11518#IT5):

| Specification   | vCPU | Memory (GB) | Throughput (pps) (received and  sent packets) | Number of queues | Private network bandwidth (Gbps)  (outbound and inbound) | CPU clock rate | Notes                       |
| --------------- | ---- | ----------- | --------------------------------------------- | ---------------- | -------------------------------------------------------- | -------------- | --------------------------- |
| IT5.16XLARGE256 | 64   | 256         | 5,000,000                                     | 16               | 23                                                       | 2.5GHz         | 4 x 3570 GB local NVMe SSDs |

2 for 21.3.13.9 and 2 for 22.3.3.44. And deploy 2 clusters with 2 shards and 0 replica.

## Test data generation
We found that the performance degrades most in distributed tables, so use SSB data to create a distributed table.

### Table creation
```
./dbgen -s 100

┌─database─┬─table────────────────┬──────rows─┬─Original size─┬─Compressed size─┐
│ default  │ lineorder_flat       │ 600037902 │ 97.07 GiB     │ 53.44 GiB       │
│ default  │ lineorder_flat_local │ 300021502 │ 48.53 GiB     │ 29.55 GiB       │
└──────────┴──────────────────────┴───────────┴───────────────┴─────────────────┘
```
* lineorder_flat: Create from standard guide, all data in one node;
* lineorder_flat_local: Create 1 distributed table with 2 shards, and then insert data from `lineorder_flat` into distributed table `lineorder_flat_all`;

### Test queries
2 set of queries, one for non-distributed table, and another for distributed table.
#### ssb_queries
All 13 standard SSB queries from the [[guide](https://clickhouse.com/docs/en/getting-started/example-datasets/star-schema/#)](https://clickhouse.com/docs/en/getting-started/example-datasets/star-schema/#).
#### dist_ssb_queries
All 13 standard SSB queries from the [guide](https://clickhouse.com/docs/en/getting-started/example-datasets/star-schema/#), but table name from **lineorder_flat** to **lineorder_flat_all**.
#### Run queries
```
# Joint test with all 13 queries.
# dist_ssb_queries for distributed table test
clickhouse-benchmark -r on --json result.json -i 5000 < [dist_]ssb_queries

# For each SSB query.
clickhouse-benchmark --json result.json -i 1000 < ssbx.x
```

## Performance
### Non-distributed table
query_time_percentiles | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 10.49 | 14.77 | 22.67 | 78.06 | 181.46 | 264.88 | 290.96 | 340.88 | 392.54 | 432.90 | 437.59 | 456.92 | 479.45 | 485.06
22.3.3.44 | 11.34 | 17.76 | 20.36 | 72.83 | 215.07 | 251.49 | 301.60 | 416.45 | 428.37 | 491.93 | 517.15 | 527.48 | 542.37 | 561.27
### Distributed table
| query_time_percentiles | 0     | 10    | 20    | 30    | 40     | 50     | 60     | 70     | 80     | 90     | 95     | 99     | 99.9   | 99.99  |
| ---------------------- | ----- | ----- | ----- | ----- | ------ | ------ | ------ | ------ | ------ | ------ | ------ | ------ | ------ | ------ |
| 21.3.13.9              | 16.19 | 20.40 | 22.83 | 68.78 | 161.13 | 181.10 | 190.94 | 261.25 | 273.65 | 300.02 | 320.80 | 329.96 | 337.44 | 353.47 |
| 22.3.3.44              | 19.57 | 25.57 | 29.37 | 94.72 | 191.79 | 202.76 | 213.08 | 271.75 | 357.13 | 398.58 | 410.79 | 443.40 | 494.78 | 513.98 |
### Details for each query(Non-distributed table)
SSB1.1 | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 55.00 | 57.35 | 58.30 | 58.90 | 59.27 | 59.77 | 60.24 | 60.64 | 61.16 | 61.92 | 62.68 | 64.68 | 80.70 | 80.70
21.12.4.1 | 62.15 | 65.21 | 66.76 | 67.82 | 68.43 | 68.76 | 69.38 | 69.78 | 70.37 | 71.26 | 72.30 | 80.45 | 103.56 | 103.56

SSB1.2 | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 12.01 | 12.40 | 12.53 | 12.62 | 12.74 | 12.86 | 12.99 | 13.10 | 13.32 | 13.75 | 14.25 | 15.99 | 19.50 | 19.50
21.12.4.1 | 15.43 | 16.34 | 16.68 | 16.87 | 17.04 | 17.12 | 17.23 | 17.38 | 17.53 | 17.79 | 17.99 | 18.74 | 19.11 | 38.75

SSB1.3 | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 10.54 | 10.87 | 10.96 | 11.03 | 11.13 | 11.21 | 11.33 | 11.46 | 11.65 | 12.52 | 12.99 | 13.48 | 32.25 | 34.35
21.12.4.1 | 10.80 | 11.35 | 11.50 | 11.60 | 11.72 | 11.82 | 11.94 | 12.06 | 12.20 | 12.69 | 13.29 | 14.21 | 32.07 | 32.07

SSB2.1 | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 279.19 | 287.22 | 288.43 | 289.21 | 289.96 | 290.86 | 291.85 | 292.61 | 293.70 | 295.65 | 297.43 | 300.80 | 303.13 | 303.13
21.12.4.1 | 296.22 | 299.60 | 300.91 | 301.92 | 302.86 | 303.76 | 304.86 | 305.92 | 307.19 | 310.27 | 312.98 | 318.23 | 320.63 | 322.97

SSB2.2 | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 269.20 | 276.88 | 278.41 | 279.34 | 280.05 | 280.89 | 281.80 | 282.68 | 283.91 | 285.68 | 286.93 | 291.40 | 294.05 | 294.05
21.12.4.1 | 244.52 | 248.97 | 250.96 | 253.15 | 259.20 | 264.10 | 266.78 | 272.08 | 296.19 | 341.02 | 368.66 | 393.20 | 396.00 | 399.93

SSB2.3 | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 231.45 | 239.29 | 241.00 | 242.06 | 243.04 | 244.04 | 245.30 | 246.52 | 247.66 | 249.86 | 252.00 | 256.26 | 257.83 | 257.83
21.12.4.1 | 238.54 | 241.60 | 243.22 | 245.56 | 250.89 | 257.24 | 260.09 | 267.81 | 297.23 | 332.76 | 359.31 | 376.91 | 382.00 | 383.38

SSB3.1 | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 426.51 | 433.62 | 436.05 | 437.29 | 438.53 | 439.68 | 440.62 | 441.97 | 443.45 | 445.43 | 447.37 | 450.55 | 457.27 | 458.78
21.12.4.1 | 497.94 | 509.13 | 512.01 | 514.51 | 516.93 | 519.10 | 520.92 | 522.85 | 525.27 | 528.52 | 531.92 | 544.53 | 552.35 | 552.35

SSB3.2 | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 379.10 | 390.17 | 392.37 | 393.87 | 394.93 | 396.22 | 397.28 | 398.68 | 399.92 | 401.65 | 402.85 | 405.29 | 411.32 | 411.32
21.12.4.1 | 464.31 | 477.87 | 481.37 | 483.72 | 485.79 | 487.27 | 488.78 | 490.56 | 492.26 | 495.38 | 497.73 | 503.57 | 509.21 | 509.43

SSB3.3 | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 322.82 | 336.82 | 338.47 | 339.72 | 340.60 | 341.57 | 342.58 | 343.69 | 344.83 | 346.50 | 348.27 | 351.78 | 355.63 | 358.10
21.12.4.1 | 394.61 | 413.64 | 416.51 | 418.33 | 420.41 | 422.51 | 424.06 | 426.27 | 429.30 | 440.91 | 497.27 | 545.32 | 554.99 | 554.99

SSB3.4 | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 12.36 | 13.15 | 13.32 | 13.47 | 13.60 | 13.72 | 13.84 | 14.03 | 14.24 | 14.59 | 15.05 | 15.82 | 31.20 | 31.20
21.12.4.1 | 17.19 | 17.96 | 18.16 | 18.32 | 18.44 | 18.59 | 18.71 | 18.86 | 19.03 | 19.29 | 19.49 | 21.31 | 31.83 | 39.32

SSB4.1 | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 426.62 | 433.32 | 435.17 | 436.38 | 437.72 | 439.15 | 440.50 | 441.76 | 443.41 | 445.80 | 447.66 | 452.32 | 455.51 | 465.10
21.12.4.1 | 413.07 | 419.10 | 420.96 | 422.15 | 423.39 | 424.59 | 425.53 | 426.83 | 428.26 | 431.02 | 432.53 | 440.01 | 447.34 | 447.34

SSB4.2 | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 163.36 | 172.32 | 173.75 | 174.80 | 175.69 | 176.35 | 176.97 | 177.79 | 178.65 | 179.65 | 180.79 | 184.06 | 188.51 | 188.51
21.12.4.1 | 200.10 | 211.67 | 213.87 | 215.28 | 216.38 | 217.39 | 218.31 | 219.50 | 220.93 | 222.89 | 224.72 | 230.45 | 234.21 | 236.14

SSB4.3 | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 152.38 | 159.34 | 161.07 | 161.78 | 162.41 | 163.21 | 163.93 | 164.57 | 165.66 | 166.81 | 167.97 | 169.36 | 170.87 | 172.84
21.12.4.1 | 197.08 | 204.86 | 207.22 | 208.76 | 209.89 | 210.91 | 211.96 | 213.13 | 214.57 | 217.27 | 219.07 | 224.26 | 225.85 | 225.85


### Details for each query(distributed table)
SSB1.1 | P0 | P10 | P20 | P30 | P40 | P50 | P60 | P70 | P80 | P90 | P95 | P99 | P99.9 | P99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 51.82 | 54.46 | 55.81 | 56.90 | 57.89 | 58.86 | 59.71 | 61.08 | 62.33 | 65.23 | 67.49 | 70.94 | 85.27 | 85.27
21.12.4.1 | 71.28 | 78.59 | 81.54 | 83.77 | 85.50 | 86.52 | 87.70 | 88.52 | 89.42 | 91.11 | 92.23 | 95.17 | 123.54 | 156.72

SSB1.2 | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 17.32 | 17.77 | 17.94 | 18.12 | 18.32 | 18.50 | 18.69 | 18.85 | 19.14 | 19.69 | 20.08 | 21.52 | 50.22 | 50.22
21.12.4.1 | 22.33 | 22.88 | 23.11 | 23.23 | 23.38 | 23.53 | 23.75 | 23.92 | 24.25 | 24.67 | 25.15 | 28.19 | 48.13 | 53.95

SSB1.3 | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 15.91 | 16.38 | 16.55 | 16.67 | 16.80 | 16.91 | 17.03 | 17.17 | 17.37 | 17.71 | 18.23 | 19.79 | 52.63 | 52.63
21.12.4.1 | 18.54 | 19.12 | 19.33 | 19.48 | 19.59 | 19.71 | 19.84 | 19.99 | 20.22 | 20.51 | 20.76 | 24.83 | 48.10 | 48.46

SSB2.1 | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 155.08 | 158.28 | 160.30 | 161.57 | 162.61 | 163.45 | 164.41 | 165.45 | 166.85 | 169.13 | 171.70 | 178.28 | 183.17 | 184.86
21.12.4.1 | 198.27 | 203.17 | 205.22 | 207.79 | 210.53 | 212.80 | 214.80 | 217.15 | 219.59 | 223.39 | 226.79 | 231.40 | 251.24 | 251.24

SSB2.2 | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 151.43 | 155.90 | 157.14 | 158.21 | 159.25 | 160.14 | 160.97 | 162.04 | 163.59 | 166.39 | 168.61 | 176.13 | 193.79 | 193.79
21.12.4.1 | 163.58 | 180.00 | 184.57 | 188.43 | 193.57 | 197.23 | 201.60 | 207.59 | 223.38 | 257.71 | 281.63 | 313.90 | 325.13 | 325.13

SSB2.3 | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 132.68 | 136.98 | 138.24 | 138.84 | 139.58 | 140.37 | 141.41 | 142.26 | 143.37 | 145.84 | 147.90 | 150.92 | 153.02 | 154.63
21.12.4.1 | 153.29 | 174.18 | 178.63 | 182.91 | 187.46 | 190.26 | 194.66 | 201.75 | 220.71 | 262.99 | 285.65 | 305.91 | 329.25 | 329.25

SSB3.1 | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 277.35 | 283.34 | 285.38 | 286.96 | 288.33 | 290.03 | 291.41 | 293.01 | 294.78 | 297.64 | 300.53 | 305.34 | 307.52 | 307.52
21.12.4.1 | 384.48 | 395.55 | 399.15 | 401.92 | 404.07 | 407.32 | 410.53 | 414.84 | 418.63 | 425.21 | 430.47 | 439.74 | 447.01 | 456.52

SSB3.2 | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 253.32 | 259.55 | 261.93 | 263.65 | 265.40 | 266.82 | 268.08 | 269.32 | 270.78 | 273.17 | 275.46 | 281.54 | 283.75 | 284.16
21.12.4.1 | 364.83 | 377.19 | 380.82 | 383.44 | 386.36 | 389.97 | 393.27 | 396.76 | 400.79 | 406.41 | 412.00 | 417.79 | 429.92 | 429.92

SSB3.3 | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 227.43 | 234.53 | 235.93 | 237.05 | 238.10 | 239.03 | 240.04 | 241.23 | 243.39 | 246.63 | 248.40 | 252.36 | 272.90 | 272.90
21.12.4.1 | 322.57 | 340.46 | 348.55 | 354.83 | 359.24 | 366.34 | 371.82 | 380.67 | 402.30 | 443.60 | 460.02 | 489.60 | 506.33 | 516.96

SSB3.4 | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 12.26 | 12.78 | 12.93 | 13.11 | 13.26 | 13.38 | 13.51 | 13.65 | 13.91 | 14.41 | 14.84 | 15.69 | 18.51 | 18.51
21.12.4.1 | 17.13 | 18.05 | 18.30 | 18.43 | 18.60 | 18.72 | 18.85 | 19.00 | 19.13 | 19.35 | 19.63 | 20.58 | 24.38 | 40.57

SSB4.1 | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 230.73 | 234.90 | 236.26 | 237.55 | 239.31 | 241.10 | 243.03 | 244.63 | 246.18 | 248.84 | 251.34 | 258.31 | 262.89 | 262.89
21.12.4.1 | 264.62 | 269.99 | 272.12 | 274.02 | 275.52 | 277.66 | 279.82 | 282.34 | 285.54 | 289.89 | 295.41 | 308.37 | 321.11 | 321.16

SSB4.2 | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 119.11 | 124.86 | 126.56 | 128.27 | 130.03 | 132.52 | 134.36 | 136.72 | 138.79 | 141.78 | 145.38 | 151.47 | 160.92 | 160.92
21.12.4.1 | 184.96 | 194.75 | 198.48 | 201.36 | 204.25 | 206.64 | 209.17 | 211.57 | 214.93 | 220.71 | 230.64 | 242.46 | 245.38 | 245.98

SSB4.3 | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 111.21 | 115.16 | 117.51 | 119.42 | 121.36 | 122.92 | 124.60 | 126.32 | 127.65 | 130.58 | 131.69 | 139.92 | 148.76 | 148.76
21.12.4.1 | 150.70 | 160.83 | 164.38 | 166.86 | 169.23 | 171.59 | 173.66 | 176.24 | 179.02 | 185.57 | 191.08 | 209.25 | 212.59 | 213.70

**Comment by alexey-milovidov** [comment]
These results are not reproduced.
Do you use the official ClickHouse builds?
If not, switching to the official ClickHouse builds should improve performance.

**Comment by chen9t** [comment]
> These results are not reproduced by another test in #36411. Do you use the official ClickHouse builds? If not, switching to the official ClickHouse builds should improve performance.

@filimonov Thanks for your reply.

Yes, It's ClickHouse builded by ourself, because we have our own deploy rules.

We are using build environment from here: [clickhouse/binary-builder](https://hub.docker.com/r/clickhouse/binary-builder), and build ClickHouse with following commands:
```
# Build
cd ${WORKSPACE}/ClickHouse
mkdir build
cd build
rm -f CMakeCache.txt
cmake ..
export THREADS=$(grep -c ^processor /proc/cpuinfo)
ninja -j $THREADS

# Output
cd ${WORKSPACE}
mkdir output
rm -rf ./output/*
mv ./ClickHouse/build/programs/clickhouse* ./output/
tar zcvf clickhouse_binary.tar.gz ./output
```
Any suggestions here?

**Comment by alexey-milovidov** [comment]
@qoega prepared an automated comparison of ClickHouse versions on several different benchmarks: Web Analytics Benchmark, Star Schema Benchmark, and Brown University Benchmark: https://clickhouse.com/benchmark/versions/

![Screenshot_20220523_003259](https://user-images.githubusercontent.com/18581488/169718828-b0ee60c9-0c2d-4ad3-8f10-8f5618655a2c.png)

22.3 is 7% faster than 21.3 and 34% faster than 20.3.

**Comment by alexey-milovidov** [comment]
@chen9t `clickhouse/binary-builder` should produce the result identical to the official binaries.
So, it is unclear what is the difference.

**Comment by chen9t** [comment]
@alexey-milovidov 

Actually, I made a comparison between [v22.3.2.2-lts](https://github.com/ClickHouse/ClickHouse/releases/tag/v22.3.2.2-lts) and [v21.3.13.9-lts](https://github.com/ClickHouse/ClickHouse/releases/tag/v21.3.13.9-lts) with standard SSB test and still got performance degradation.:

query_time_percentiles | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 9.33 | 14.07 | 21.84 | 78.06 | 188.43 | 276.08 | 295.89 | 361.92 | 420.24 | 446.79 | 459.93 | 474.22 | 482.77 | 489.37
22.3.2.2 | 10.65 | 16.09 | 19.00 | 73.63 | 225.89 | 252.88 | 296.71 | 426.79 | 458.13 | 533.16 | 553.83 | 570.09 | 579.45 | 593.36
Improvement | -12.34% | -12.57% | 14.95% | 6.01% | -16.58% | 9.18% | -0.28% | -15.20% | -8.27% | -16.20% | -16.96% | -16.82% | -16.69% | -17.53%

**Comment by chen9t** [comment]
Found one potential issue between 2 versions from trace log:

In [v21.3.13.9-lts](https://github.com/ClickHouse/ClickHouse/releases/tag/v21.3.13.9-lts) all where conditions will be moved to prewhere, Q3.3 for example(but not only for Q3.3):
```
2022.05.24 10:55:25.609470 [ 51186 ] {ae816427-1764-4934-a787-d83005086b9d} <Debug> executeQuery: (from 127.0.0.1:38341, user: clickhouse_admin, using production parser) SELECT C_CITY, S_CITY, toYear(LO_ORDERDATE) AS year, sum(LO_REVENUE) AS revenue FROM lineorder_flat WHERE (C_CITY = 'UNITED KI1' OR C_CITY = 'UNITED KI5') AND (S_CITY = 'UNITED KI1' OR S_CITY = 'UNITED KI5') AND year >= 1992 AND year <= 1997 GROUP BY C_CITY, S_CITY, year ORDER BY year ASC, revenue DESC;
2022.05.24 10:55:25.610163 [ 51186 ] {ae816427-1764-4934-a787-d83005086b9d} <Debug> InterpreterSelectQuery: MergeTreeWhereOptimizer: condition "((toYear(LO_ORDERDATE) AS year) >= 1992) AND (year <= 1997) AND ((C_CITY = 'UNITED KI1') OR (C_CITY = 'UNITED KI5')) AND ((S_CITY = 'UNITED KI1') OR (S_CITY = 'UNITED KI5'))" moved to PREWHERE
```

In [v21.3.13.9-lts](https://github.com/ClickHouse/ClickHouse/releases/tag/v21.3.13.9-lts), only some conditions in where will be moved to prewhere:
```
2022.05.24 10:55:20.343876 [ 15123 ] {c4cd99be-e72e-4946-a71f-ec5899486bb0} <Debug> executeQuery: (from 127.0.0.1:36368, user: clickhouse_admin)  SELECT C_CITY, S_CITY, toYear(LO_ORDERDATE) AS year, sum(LO_REVENUE) AS revenue FROM lineorder_flat WHERE (C_CITY = 'UNITED KI1' OR C_CITY = 'UNITED KI5') AND (S_CITY = 'UNITED KI1' OR S_CITY = 'UNITED KI5') AND year >= 1992 AND year <= 1997 GROUP BY C_CITY, S_CITY, year ORDER BY year ASC, revenue DESC;
2022.05.24 10:55:20.344560 [ 15123 ] {c4cd99be-e72e-4946-a71f-ec5899486bb0} <Debug> InterpreterSelectQuery: MergeTreeWhereOptimizer: condition "((toYear(LO_ORDERDATE) AS year) >= 1992) AND (year <= 1997)" moved to PREWHERE
2022.05.24 10:55:20.345176 [ 15123 ] {c4cd99be-e72e-4946-a71f-ec5899486bb0} <Debug> InterpreterSelectQuery: MergeTreeWhereOptimizer: condition "(toYear(LO_ORDERDATE) >= 1992) AND (toYear(LO_ORDERDATE) <= 1997)" moved to PREWHERE
```

So I changed all WHERE into PREWHERE in SSB tests, and here's the result:

query_time_percentiles | 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 95 | 99 | 99.9 | 99.99
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
21.3.13.9 | 9.33 | 14.07 | 21.84 | 78.06 | 188.43 | 276.08 | 295.89 | 361.92 | 420.24 | 446.79 | 459.93 | 474.22 | 482.77 | 489.37
22.3.2.2 | 8.71 | 14.45 | 17.86 | 176.10 | 194.24 | 247.80 | 292.14 | 340.46 | 419.06 | 448.27 | 453.49 | 459.68 | 466.13 | 501.57
Improvement | 6.65% | -2.69% | 18.24% | -125.60% | -3.08% | 10.24% | 1.27% | 5.93% | 0.28% | -0.33% | 1.40% | 3.07% | 3.45% | -2.49%


Any ideas here?

**Comment by alexey-milovidov** [comment]
Ok, now we understand that it is not related to compression,
but is related to PREWHERE optimization (btw, it's the only cost-based optimizationr that we have in ClickHouse).

**Comment by alexey-milovidov** [comment]
@chen9t why closed?

**Comment by davenger** [comment]
I found this fix #28139 made in Aug 2021 that could make fewer conditions be moved from WHERE to PREWHERE. I tried to revert this change I indeed see that it changes what conditions are moved to PREWHERE in q3.3

**Comment by alexey-milovidov** [comment]
Version 22.3 is no longer supported; you need to update to the latest version.
